py-rse/style.Rmd

```diff
@@ -4,207 +4,255 @@
+              for col in range(original.numCol):
+               result[row, col] = original[col, row]
+  return result
+```
 
-Look for algorithmic problems first.
-:   Code review isn't just (or even primarily) about style:
-    its real purpose is to find bugs before they can affect anyone.
-    The first pass over any change should therefore look for algorithmic problems.
-    Are the calculations right?
-    Are any rare cases going to be missed?
-    Are errors being caught and handled Chapter \@ref(logging)?
+However,
+"legal" doesn't mean "readable".
+That same code is easier for the eye to follow
+when nested blocks step in the same amount each time:
 
-Use a rubric.
-:   Linters are great,
-    but can't decide when someone should have used a lookup table instead of conditionals.
-    A list of things to check for can make review faster and more comprehensible,
-    especially when you can copy-and-paste or drag-and-drop specific comments
-    onto specific lines
-    (something that GitHub unfortunately doesn't yet support).
+```python
+def transpose(original):
+    result = Matrix(original.numRow, original.numCol)
+    for row in range(original.numRow):
+        for col in range(original.numCol):
+            result[row, col] = original[col, row]
+    return result
+```
 
-Ask for clarification.
-:   If you don't understand something,
-    or don't understand why the author did it,
-    ask.
-    (And when the author explains it,
-    think about suggesting that the explanation ought to be documented somewhere.)
+The use of 4 spaces is a compromise between 2
+(which we find perfectly readable, but some people find too crowded)
+and 8
+(which most people agree uses up too much horizontal space).
+As for the use of spaces rather than tabs,
+the original reason was that the most common interpretation of tabs
+by the editors of the 1980s was 8 spaces,
+which again was more than most people felt necessary.
+Today,
+almost all editors will auto-indent or auto-complete when the tab key is pressed,
+but the legacy of those times lives on.
+
+**Do not put spaces inside parentheses.**
+
+Don't write <code>(&nbsp;1+2&nbsp;)</code>.
+This applies to function calls as well:
+write `max(a, b)` rather than <code>max(&nbsp;a,&nbsp;b&nbsp;)</code>.
 
-Offer alternatives.
-:   Telling authors that something is wrong is helpful;
-    telling them what they might do instead is more so.
+**Always use spaces around comparisons like `>` and `<=`.**
 
-Don't be sarcastic or disparaging.
-:   "Did you maybe think about *testing* this garbage?"
-    is a Code of Conduct violation in any well-run project.
+Python automatically interprets `a+b<c+d` as `(a+b)<(c+d)`,
+but that's a lot of punctuation crowded together;
+using spaces around comparison operators makes it easier to see
+what's being compared to what.
+However,
+you should use your own judgment for spacing around arithmetic operators like `+` and `/`.
+For example,
+`a+b+c` is perfectly readable,
```

Comment:
Perhaps mention that spacing can effectively be used to indicate priority without having to clutter with parenthesis? As in `a*b + c`.

Reply:
I think this is bad style: `a*b + c*d` is OK, but what how do you add spaces to `a**2*b+c**2*d`? Safer to tell people to parenthesize...

Reply:
I meant spaces as a complement to parenthesis. In simple examples, only spaces offers more readability than parenthesis in my opinion, and in more complex cases both can be used so it doesn't get cluttered with too many parentheses. If I wanted to clarify the expression you gave, I would write it as 

```
(a**2)*b + (c**2)*d
```

which I think is clearer than both 
```
(a**2)*b+(c**2)*d
```

and

```
((a**2)*b)+((c**2)*d)
```

Reply:
PEP-8 doesn't appear to mandate anything on this, and since we differ on what to recommend, I feel we should leave this as is.

Reply:
This is actually in PEP 8, third bullet under [Other Recommendations](https://www.python.org/dev/peps/pep-0008/#other-recommendations):

> If operators with different priorities are used, consider adding whitespace around the operators with the lowest priority(ies). Use your own judgment; however, never use more than one space, and always have the same amount of whitespace on both sides of a binary operator:
>
> ```python
> # Correct:
> i = i + 1
> submitted += 1
> x = x*2 - 1
> hypot2 = x*x + y*y
> c = (a+b) * (a-b)
> ```
> 
> ```python
> # Wrong:
> i=i+1
> submitted +=1
> x = x * 2 - 1
> hypot2 = x * x + y * y
> c = (a + b) * (a - b)
> ```

A gotcha might worth mentioning is that while whitespace conveys priority to the reader, it does not convey it to the Python interpreter so it can't be treated the same as surrounding parentheses.